### DISCUSSION TITLE
Add-Type vs using assembly for TSQL ScriptDOM

### DISCUSSION THREAD

**Discussion by ronascentes** [PowerShell/PowerShell#18421] [Q&A] [answered]
**TLDR:** Can someone please clarify why I have to use `Add-Yype` to add TSQL ScriptDOM assembly and not `using assembly`? 

Must use: 
`Add-Type -Path "Microsoft.SqlServer.TransactSql.ScriptDom.dll"`

But cannot use:
`using assembly 'Microsoft.SqlServer.TransactSql.ScriptDom.dll'`
`using namespace Microsoft.SqlServer.TransactSql.ScriptDom`

**Long Description:** I'm creating a new class that derivates from either TSqlConcreteFragmentVisitor from the Microsoft.SqlServer.TransactSql.ScriptDom namespace:

```
class VisitorSelectStar: Microsoft.SqlServer.TransactSql.ScriptDom.TSqlConcreteFragmentVisitor 
{
    [void]Visit ([Microsoft.SqlServer.TransactSql.ScriptDom.SelectStarExpression] $fragment) 
    {
        $errorline = "WARNING: 'SELECT * found at "
        Write-Host $errorline $fragment.StartLine ":" $fragment.StartColumn ":" $fragment.FragmentLength
    }
}
```
Since my class that derives from a type defined in an external assembly will not compile unless the assembly with the base type has already been loaded. Consequently, the Microsoft.SqlServer.TransactSql.ScriptDom.dll assembly must first be loaded before a script with a VisitorSelectStar class can be run.
`Add-Type -Path "Microsoft.SqlServer.TransactSql.ScriptDom.dll"`

The problem, which was already ponted by [Dan Guzmon](https://www.dbdelta.com/microsoft-sql-server-script-dom/) is I cannot run `Add-Type` interactively. So a wrapper script can be employed to execute the `Add-Type` command and then “[dot source](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scripts)” the script file with the VisitorSelectStar class. 

But according to the [PowerShell Documentation](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_using?view=powershell-7.2), I could use `using assembly` to preload types from a .NET assembly into a script at parse time what would allow me to create new PowerShell classes that use types from the preloaded assembly. But when I try use it instead of `Add-Type` my script fails:

```
Line |
   4 |  … SelectStar: Microsoft.SqlServer.TransactSql.ScriptDom.TSqlConcreteFra …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to find type [Microsoft.SqlServer.TransactSql.ScriptDom.TSqlConcreteFragmentVisitor].
```

Why I cannot use `using assembly` and `using namespace` for Microsoft.SqlServer.TransactSql.ScriptDom.dll instead of `Add-Type`? What's the difference between `Add-Type` and `using assembly`?

**Comment by ronascentes** [comment, marked as answer]
It seems a by-design behavior. The assembly is not loaded at parse time, but instead at compile time. There is a lot of issues reporting this same behavior ([#6652](https://github.com/PowerShell/PowerShell/issues/6652), [#3641](https://github.com/PowerShell/PowerShell/issues/3641), [#2074](https://github.com/PowerShell/PowerShell/issues/2074), [#11856](https://github.com/PowerShell/PowerShell/issues/11856), [#12547](https://github.com/PowerShell/PowerShell/issues/12547))
